### ISSUE TITLE
Implement KNN Models for Options IV Prediction with Feature Engineering

### ISSUE THREAD

**Issue by kBrutal** [kBrutal/Dynamic-MultiAgent-System-for-Accessing-Research-papers#1]
## Task Description
Implement multiple KNN regression models to predict implied volatility (IV) values for put and call options across different strike prices.

### Data Overview
We have a dataset containing:
- Timestamps (epoch format)
- Underlying price
- Expiry dates
- Multiple IV values for calls (strikes 23500-26000)
- Multiple IV values for puts (strikes 22500-25000)
- Additional features X0-X41

### Required Implementation Steps

1. **Data Preprocessing**
   - Convert epoch timestamp to datetime format
   - Calculate Time to Expiry feature using expiry and timestamp
   - Create separate dataframes for put and call IVs
   - Structure features: strike price, put/call indicator

2. **Feature Engineering**
   - Distance from current price to strike (moneyness)
   - Time decay features
   - Volatility surface related features
   - Include existing X0-X41 features

3. **Model Implementation**
```python
# Sample code structure
from sklearn.neighbors import KNeighborsRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
import numpy as np

def create_features(df):
    # Convert timestamp
    df['datetime'] = pd.to_datetime(df['timestamp'], unit='s')
    df['expiry'] = pd.to_datetime(df['expiry'])
    df['time_to_expiry'] = (df['expiry'] - df['datetime']).dt.total_seconds() / (24*60*60)
    
    # Additional features
    df['moneyness'] = df['strike'] / df['underlying']
    return df

def train_knn_models(X, y_columns):
    results = {}
    for col in y_columns:
        # Train-test split
        X_train, X_test, y_train, y_test = train_test_split(
            X, df[col], test_size=0.2, random_state=42
        )
        
        # Train model
        knn = KNeighborsRegressor(n_neighbors=5)
        knn.fit(X_train, y_train)
        
        # Predictions
        y_pred = knn.predict(X_test)
        
        # Metrics
        results[col] = {
            'rmse': np.sqrt(mean_squared_error(y_test, y_pred)),
            'mse': mean_squared_error(y_test, y_pred),
            'mae': mean_absolute_error(y_test, y_pred)
        }
    
    return results
```

4. **Evaluation Requirements**
   - Use train-test split (80-20)
   - Calculate for each IV prediction:
     - RMSE (Root Mean Squared Error)
     - MSE (Mean Squared Error)
     - MAE (Mean Absolute Error)
   - No standardization/normalization of target variables

### Important Notes
1. Create separate models for each strike price IV
2. Handle put and call IVs separately
3. No preprocessing of target variables required
4. Ensure proper feature engineering for improved predictions
5. Document metrics for all predicted columns

### Deliverables
1. Preprocessed datasets for puts and calls
2. Feature engineering implementation
3. KNN models for each IV target
4. Comprehensive metrics report
5. Code documentation and comments
